buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectGenerate.java

```diff
@@ -0,0 +1,403 @@
+    this.outputFile = outputFile;
+  }
+
+  /**
+   * Extracts functions and generates parserImpls.ftl for the given dialect.
+   */
+  public void run() {
+    Map<String, String> functions = extractFunctions();
+    for (Map.Entry<String, String> entry : functions.entrySet()) {
+      String key = entry.getKey();
+      String value = entry.getValue();
+      System.out.println(key + "=" + value + "\n");
+    }
+  }
+
+  /**
+   * Traverses the parsing directory structure and extracts all of the
+   * functions located in *.ftl files into a Map.
+   */
+  public Map<String, String> extractFunctions() {
+    Queue<String> queue = getTraversalPath(rootDirectory);
+    Map<String, String> functionMap = new LinkedHashMap<String, String>();
+    traverse(queue, rootDirectory, functionMap);
+    return functionMap;
+  }
+
+  /**
+   * Generates the parserImpls.ftl file for the dialect.
+   *
+   * @param functions The functions to place into the output file
+   */
+  public void generateParserImpls(Map<String, String> functions) {}
+
+  /**
+   * Gets the traversal path for the dialect by "subtracting" the root
+   * absolute path from the dialect directory absolute path.
+   *
+   * @param rootDirectoryFile The file for the root parsing directory
+   */
+  private Queue<String> getTraversalPath(File rootDirectoryFile) {
+    String dialectPath = dialectDirectory.getAbsolutePath();
+    String rootPath = rootDirectory.getAbsolutePath();
+    int rootIndex = dialectPath.indexOf(rootPath);
+    dialectPath = dialectPath.substring(rootIndex + rootPath.length() + 1);
+    return new LinkedList(Arrays.asList(dialectPath.split("/")));
+  }
+
+  /**
+   * Traverses the determined path given by the queue. Once the queue is
+   * empty, the dialect directory has been reached. In that case any *.ftl
+   * file should be processed and no further traversal should happen.
+   *
+   * @param directories The directories to traverse in topdown order
+   * @param currentDirectory The current directory the function is processing
+   * @param functionMap The map to which the parsing functions will be added to
+   */
+  private void traverse(
+      Queue<String> directories,
+      File currentDirectory,
+      Map<String, String> functionMap) {
+    File[] files = currentDirectory.listFiles();
+    // Ensures that files are processed first.
+    Arrays.sort(files, new Comparator<File>() {
+      @Override
+      public int compare(File file1, File file2) {
+        return Boolean.compare(file1.isDirectory(), file2.isDirectory());
+      }
+    });
+    String nextDirectory = directories.peek();
+    for (File f : files) {
+      String extension = "";
+      String fileName = f.getName();
+      int i = fileName.lastIndexOf('.');
+      if (i > 0) {
+        extension = fileName.substring(i+1);
+      }
+      if (f.isFile() && extension.equals("ftl")) {
+        try {
+          processFile(f, functionMap);
+        } catch(IOException e) {
```

Comment:
Is IOException the only exception that could be thrown in this call? It looks like IllegalStateException is also an option.